mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -276,44 +277,53 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
       num: Int): RDD[(Int, Array[(Int, Double)])] = {
     val srcBlocks = blockify(rank, srcFeatures)
     val dstBlocks = blockify(rank, dstFeatures)
-    val ratings = srcBlocks.cartesian(dstBlocks).flatMap {
-      case ((srcIds, srcFactors), (dstIds, dstFactors)) =>
-        val m = srcIds.length
-        val n = dstIds.length
-        val ratings = srcFactors.transpose.multiply(dstFactors)
-        val output = new Array[(Int, (Int, Double))](m * n)
-        var k = 0
-        ratings.foreachActive { (i, j, r) =>
-          output(k) = (srcIds(i), (dstIds(j), r))
-          k += 1
+    /**
+     * Use dot to replace blas 3 gemm is the key approach to improve efficiency.
```

Comment:
I think we should say something like:

> The previous approach used for computing top-k recommendations aimed to group individual factor vectors into blocks, so that Level 3 BLAS operations (gemm) could be used for efficiency. However, this causes excessive GC pressure due to the large arrays required for intermediate result storage, as well as a high sensitivity to the block size used.
>
> The following approach still groups factors into blocks, but instead computes the top-k elements per block, using Level 1 BLAS (dot) and an efficient BoundedPriorityQueue. This avoids any large intermediate datastructures and results in significantly reduced GC pressure as well as shuffle data, which far outweighs any cost incurred from not using Level 3 BLAS operations.